058-kraft-upgrades-and-downgrades.md

```diff
@@ -0,0 +1,221 @@
+1. Downgrading the `metadata.version`
+2. Rolling out the older Kafka version
+
+Changes to the metadata formats - that are defined by the `metadata.version` - might not be backwards compatible.
+As a result, the first step of the downgrade procedure - downgrading the `metadata.version` - can be done safely without any metadata loss only for selected metadata versions.
+For other versions, the downgrade will be possible only as an _unsafe_ downgrade that might result in losing some metadata information, which might have a negative impact on the cluster.
+
+#### Existing downgrade limitations
+
+As of today, the Kraft downgrade situation is as follows:
+1. The unsafe downgrade is currently not supported by Apache Kafka.
+2. None of the `metadata.versions` supported by current Kafka versions support the safe downgrade.
+   _(You should be able to downgrade from 3.5-IV1 to 3.5-IV0. But you cannot downgrade between 3.5 and 3.4 metadata versions.)_
+
+The first point should be addressed in the future when the support for the unsafe downgrade is implemented.
+The second point is expected to remain an issue.
+While in the future the metadata formats might become more stable and it might be possible to safely downgrade the metadata between different Kafka versions, there still might be some versions that do not support the safe downgrade.
+
+Despite Kafka currently not really supporting downgrades, this proposal designs the Strimzi implementation based on how the downgrade is expected to work.
+
+### Strimzi implementation
+
+Strimzi will follow the same approach to upgrading / downgrading a KRaft cluster as it does today for ZooKeeper-based clusters.
+
+#### Configuring the metadata version
+
+In ZooKeeper-based clusters, the `inter.broker.protocol.version` is stored in the Kafka configuration (`Kafka.spec.kafka.config`).
+The `metadata.version` is not part of the Kafka configuration and cannot be managed declaratively.
+To allow users to configure it in our Kafka custom resource, a new String-type field named `metadataVersion` will be added to `Kafka.spec.kafka`.
+This field will be validated by the operator to contain a valid Kafka metadata version.
+
+Users can use this new field to configure the metadata version.
+The default value of this field when not set will be the metadata version corresponding to the current Kafka version.
+The default version corresponding to a given Kafka version will be stored in `kafka-versions.yaml` in the same way we today store the default `inter.broker.protocol.version` and `log.message.format.version`.
+
+The operator will use this field to set the metadata version in the Kafka cluster.
+In the initial deployment of the Kafka cluster, it will be set using the `kafka-storage.sh` tool.
+In existing Kafka clusters, it will be queried and changed using the Kafka Admin API.
+
+The current metadata version will be tracked in the `.status` section of the Kafka CR in `.status.kafkaMetadataVersion`.
```

Comment:
Good point. I already see new features coming in from recent KIPs.

Reply:
Kafka does not make it clear what the intentions are. There is no clear documentation of how the features should be used or even a list of them beyond the `metadata.version`. It is also hard to predict what impact might the other features have for upgrades / downgrades and how will they integrate with the `metadata.version`. 

This proposal covers only the `metadata.version` as currently the only feature relevant to upgrades / downgrades and managed by Strimzi. It also designs the `.spec` and `.status` sections accordingly. There is currently no reason to use some complicated maps for them instead of a simple field.

The expectation is, that after ZooKeeper support is dropped, we will migrate to a new API version that will give us the chance to re-evaluate this decision if needed.